File: ios/SQLiteBatchCore.m
#import "SQLiteBatchCore.h"

#include "sqlite-connection-core.h"

@implementation SQLiteBatchCore

static int bindObjectValue(int connection_id, NSObject * value, int index)
{
  int prepareResult;

  if (value == nil) {
    prepareResult = scc_bind_null(connection_id, index);
  } else if ([value isKindOfClass: [NSNumber class]]) {
    prepareResult =
      scc_bind_double(connection_id, index, [(NSNumber *)value doubleValue]);
  } else if ([value isKindOfClass: [NSString class]]) {
    prepareResult =
      scc_bind_text(connection_id, index, [(NSString *)value cString]);
  } else {
    prepareResult = scc_bind_null(connection_id, index);
  }

  return prepareResult;
}

+ (void) initialize
{
  scc_init();
}

+ (int) openBatchConnection: (NSString *) fullName
                      flags: (int) flags
{
  return scc_open_connection([fullName cString], flags);
}

+ (NSArray *) executeBatch: (int) connection_id
                      data: (NSArray *) data
{
  NSMutableArray * results = [NSMutableArray arrayWithCapacity: 0];

  for (int i=0; i < [data count]; ++i) {
    int previousTotalChanges = scc_get_total_changes(connection_id);

    NSArray * entry = [data objectAtIndex:i];

    NSString * statement = [entry objectAtIndex: 0];

    int prepareResult =
      scc_begin_statement(connection_id, [statement cString]);

    if (prepareResult != 0) {
      // do nothing here; error will be handled below
    } else if ([[entry objectAtIndex: 1] isKindOfClass: [NSDictionary class]]) {
      NSDictionary * parameters = [entry objectAtIndex: 1];

      for (NSString * key in parameters) {
        // NOTE: zero parameter index will result in a bind error below.
        int parameter_index =
          scc_bind_parameter_index(connection_id, [key cString]);

        NSObject * bindValue = [parameters valueForKey: key];

        prepareResult =
          bindObjectValue(connection_id, bindValue, parameter_index);

	// break here & report error in case a named parameter is not found
        if (prepareResult != 0) break;
      }
    } else {
      NSArray * bind = [entry objectAtIndex: 1];

      for (int j=0; j < [bind count]; ++j) {
        NSObject * bindValue = [bind objectAtIndex: j];

        prepareResult = bindObjectValue(connection_id, bindValue, 1 + j);

        // should only break if there are too many parameters here
        if (prepareResult != 0) break;
      }
    }

    if (prepareResult != 0) {
      [results addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];

      scc_end_statement(connection_id);

      continue;
    }

    int stepResult = scc_step(connection_id);

    if (stepResult == 100) {
      const int columnCount = scc_get_column_count(connection_id);

      NSMutableArray * columns = [NSMutableArray arrayWithCapacity: 0];

      for (int j = 0; j < columnCount; ++j) {
        NSString * columnNameAsString =
          [NSString stringWithUTF8String: scc_get_column_name(connection_id, j)];
        [columns addObject: columnNameAsString];
      }

      NSMutableArray * rows = [NSMutableArray arrayWithCapacity: 0];

      do {
        NSMutableArray * row = [NSMutableArray arrayWithCapacity: 0];

        for (int j = 0; j < columnCount; ++j) {
          const int columnType = scc_get_column_type(connection_id, j);

          if (columnType == SCC_COLUMN_TYPE_INTEGER ||
              columnType == SCC_COLUMN_TYPE_FLOAT) {
            NSNumber * columnNumberValue =
              [NSNumber numberWithDouble: scc_get_column_double(connection_id, j)];
            [row addObject: columnNumberValue];
          } else if (columnType == SCC_COLUMN_TYPE_NULL) {
            [row addObject: [NSNull null]];
          } else {
            NSString * columnStringValue =
              [NSString stringWithUTF8String: scc_get_column_text(connection_id, j)];
            [row addObject: columnStringValue];
          }
        }

        [rows addObject: row];

        stepResult = scc_step(connection_id);
      } while(stepResult == 100);

      [results addObject: @{@"status":@0, @"columns": columns, @"rows": rows}];
    } else if (stepResult == 101) {
      int totalChanges = scc_get_total_changes(connection_id);
      int rowsAffected = totalChanges - previousTotalChanges;

      [results addObject: @{
        @"status": @0,
        @"rowsAffected": [NSNumber numberWithInteger: rowsAffected],
        @"totalChanges": [NSNumber numberWithInteger: totalChanges],
        @"lastInsertRowId": [NSNumber numberWithInteger:
          scc_get_last_insert_rowid(connection_id)]
      }];
    } else {
      [results addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];
    }

    scc_end_statement(connection_id);
  }

  return results;
}

@end

File: android/src/main/java/io/sqlc/SQLiteBatchCore.java
package io.sqlc;

import org.json.JSONArray;
import org.json.JSONObject;

public class SQLiteBatchCore {
  static public int openBatchConnection(String fullName, int flags) {
    return SCCoreGlue.scc_open_connection(fullName, flags);
  }

  static public JSONArray executeBatch(int mydbc, JSONArray data) {
    try {
      final int count = data.length();

      JSONArray results = new JSONArray();

      for (int i=0; i<count; ++i) {
        int previousTotalChanges = SCCoreGlue.scc_get_total_changes(mydbc);

        JSONArray entry = data.getJSONArray(i);

        String s = entry.getString(0);

        if (SCCoreGlue.scc_begin_statement(mydbc, s) != 0) {
          JSONObject result = new JSONObject();
          result.put("status", 1); // REPORT SQLite ERROR 1
          result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
          results.put(result);
        } else {
          int bindResult = 0; // SQLite OK

          if ((entry.get(1) instanceof JSONArray)) {
            final JSONArray bind = entry.getJSONArray(1);
            final int bindCount = bind.length();

            for (int j = 0; j < bindCount; ++j) {
              final Object o = bind.get(j);

              bindResult = bindObjectValue(mydbc, o, 1 + j);

              // should only break if there are too many parameters here
              if (bindResult != 0) break;
            }
          } else {
            final JSONObject bind = entry.getJSONObject(1);
            final int bindCount = bind.length();
            final JSONArray names = bind.names();

            for (int j = 0; j < bindCount; ++j) {
              final String name = names.getString(j);
              final Object o = bind.get(name);
              final int bindIndex =
                SCCoreGlue.scc_bind_parameter_index(mydbc, name);

              bindResult = bindObjectValue(mydbc, o, bindIndex);

              // break here & report error if a named parameter is not found
              if (bindResult != 0) break;
            }
          }

          if (bindResult != 0) {
            JSONObject result = new JSONObject();
            result.put("status", 1); // REPORT SQLite ERROR 1
            result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
            continue;
          }

          int stepResult = SCCoreGlue.scc_step(mydbc);

          if (stepResult == 100) {
            final int columnCount = SCCoreGlue.scc_get_column_count(mydbc);

            JSONArray columns = new JSONArray();

            for (int j=0; j < columnCount; ++j) {
              columns.put(SCCoreGlue.scc_get_column_name(mydbc, j));
            }

            JSONArray rows = new JSONArray();

            do {
              JSONArray row = new JSONArray();

              for (int col=0; col < columnCount; ++col) {
                final int type = SCCoreGlue.scc_get_column_type(mydbc, col);

                if (type == SCCoreGlue.SCC_COLUMN_TYPE_INTEGER ||
                    type == SCCoreGlue.SCC_COLUMN_TYPE_FLOAT) {
                  row.put(SCCoreGlue.scc_get_column_double(mydbc, col));
                } else if (type == SCCoreGlue.SCC_COLUMN_TYPE_NULL) {
                  row.put(JSONObject.NULL);
                } else {
                  row.put(SCCoreGlue.scc_get_column_text(mydbc, col));
                }
              }

              rows.put(row);

              stepResult = SCCoreGlue.scc_step(mydbc);
            } while (stepResult == 100);

            JSONObject result = new JSONObject();
            result.put("status", 0); // REPORT SQLite OK
            result.put("columns", columns);
            result.put("rows", rows);
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          } else if (stepResult == 101) {
            int totalChanges = SCCoreGlue.scc_get_total_changes(mydbc);
            int rowsAffected = totalChanges - previousTotalChanges;

            JSONObject result = new JSONObject();
            // same order as iOS & macOS ("osx"):
            result.put("status", 0); // REPORT SQLite OK
            result.put("totalChanges", totalChanges);
            result.put("rowsAffected", rowsAffected);
            result.put("lastInsertRowId",
              SCCoreGlue.scc_get_last_insert_rowid(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          } else {
            JSONObject result = new JSONObject();
            result.put("status", 1); // REPORT SQLite ERROR 1
            result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          }
        }
      }

      return results;
    } catch(Exception e) {
      // NOT EXPECTED - internal error:
      throw new RuntimeException(e);
    }
  }

  static private int bindObjectValue(int mydbc, Object value, int index)
  {
    int bindResult = 0; // SQLite OK

    if (value instanceof Number) {
      bindResult =
        SCCoreGlue.scc_bind_double(mydbc, index, ((Number)value).doubleValue());
    } else if (value instanceof String) {
      bindResult = SCCoreGlue.scc_bind_text(mydbc, index, value.toString());
    } else {
      bindResult = SCCoreGlue.scc_bind_null(mydbc, index);
    }
    return bindResult;
  }

  static {
    System.loadLibrary("sqlc-connection-core-glue");
    SCCoreGlue.scc_init();
  }
}
